staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -503,6 +503,11 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
 		<-delayedStopCh.Signaled()
 
+		// Additionally wait for preshutdown hooks to also be finished, as some of them need
+		// to send API calls to clean up after themselves (e.g. lease reconcilers removing
+		// itself from the active servers).
+		<-preShutdownHooksHasStoppedCh
```

Comment:
yeah, I think this is okay

Reply:
why do we want to block twice on preShutdownHooks? 

https://github.com/kubernetes/kubernetes/pull/110000/files#diff-68eb390e148644c38174960be4d103c0e377e71eb540aa8c395f7d2278363973R479-R485

Reply:
It's not blocking twice - it's blocking different thing:
1) here we're ensuring that we won't "activate" the waitgroup filter (which is effectively rejecting with RetryAfter all incoming requests)
2) the one you linked above is ensuring that preshutdown hooks will block closing the HTTP server closure

(1) [what I'm adding here] should happen earlier.
So what you're suggesting is that the one that is existing is no longer needed?

Reply:
not suggesting, it was curiosity about the possible consequences of the changes, there are many goroutines and channels here and I'm afraid we deadlock somehow ... 
how hard is to unit test this change? https://github.com/kubernetes/kubernetes/blob/076168b84d0af4ad65cb5664fc1cef40f837e9dc/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go#L431

Reply:
The existing tests are failing - which is good actually :)

So yes - we need to modify them to actually test what we need (which I guess is roughly what you're asking for - because we need to cover this usecase).

I will work on it in upcoming days (though not today).

Reply:
@aojea - I added a diagram above to model what waits on what: first commit the status quo, in the second (main commit) I'm just updating it with the addition that I'm adding.
As I mentioned - this seems to be fine to me.

Regarding the unit test:
- the failing one: `TestPreShutdownHooks` doesn't make any sense to me - in fact I would say it's wrong (it doesn't make any sense to register the hook that will be waiting for draining requests, because we have a dedicated logic for that). And as motivated by this PR, we actually want the opposite - we want to ensure that we won't start draining before preshudown hooks are done.

I will update the test to test the opposite direction if we agree on the above ^^

Reply:
I can't remember the details, but that change was required because it was closing the Listener .... with the Listener closed, the port is closed, not new requests ...

Reply:
it was closing one of these channels  before the preshutdown hooks

https://github.com/kubernetes/kubernetes/blob/a84c1b71005930e8253c1348515020132c5c175b/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L444

closing the listener, so no new request can be created because the port was not listening

Reply:
But closing listener before preshutdown hooks is incorrect - becasue preshutdown hooks may need to send API calls (and some do send).

In real prod scenario where we have this default 70s `ShutdownDelayDuration`:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/lifecycle_signals.go#L42
preshutdown hooks are alrady finished when we start rejecting new requests.
But in tests (where we don't use these 70s) - we actually need to ensure that preshutdown hooks had a chance to run.

Reply:
`TestPreShutdownHooks` seems to be broken - I have made some changes that make it functional I believe, check https://github.com/kubernetes/kubernetes/pull/110026

please note that there are two modes graceful shutdown, the new mode where we don't shutdown the HTTP server until all requests in flight drain can be enabled by `shutdown-send-retry-after`.

I think we are discussing the original graceful termination approach here.

> But closing listener before preshutdown hooks is incorrect - becasue preshutdown hooks may need to send API calls (and some do send).

yes, that's correct, I think i would add to it by saying that preshutdown hooks should be invoked only after requests in flight are drained because in-flight requests might depend on those states cleaned up by the preshutdown hooks.

Reply:
> But closing listener before preshutdown hooks is incorrect - becasue preshutdown hooks may need to send API calls (and some do send).

that was the bug fixed by https://github.com/kubernetes/kubernetes/pull/108033 :)

> the current sequence roughly is:
> 
> * T0: ShutdownInitiated: TERM signal received
>   
>   * run pre shutdown hooks
>   * /readyz starts returning red
> * T0+ShutdownDelayDuration:
> * wait for preshutdown hooks to finish (this PR accomplishes this)
> * reject incoming request(s) here on by invoking `HandlerChainWaitGroup.Wait()`
> * initiate net HTTP server shutdown
> * wait for requests in flight to drain (this includes requests from preshutdown hook

the last sentence is confusing, since you mention that you've already waited "for preshutdown hooks to finish"